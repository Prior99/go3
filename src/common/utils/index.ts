export * from "./board-color";
export * from "./board-sizes";
export * from "./encrypt";
export * from "./environment";
export * from "./is-browser";
export * from "./rating";
export * from "./rank";
export * from "./authorization";
