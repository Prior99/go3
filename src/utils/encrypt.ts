import { createHmac } from "crypto";

const SECRET = "Eech3ahm";

export function hash(value: string) {
    return createHmac("sha256", SECRET).update(value).digest("hex");
}
