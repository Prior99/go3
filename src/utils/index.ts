export * from "./environment";
export * from "./require-login";
