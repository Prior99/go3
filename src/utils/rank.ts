export enum RankClass {
    KYU = "kyu",
    DAN = "dan",
}

export class Rank {
    public rank: number;
    public rankClass: RankClass;
    public rating: number;

    constructor(rating: number) {
        this.rating = rating;
        if (rating < 2100) {
            this.rank = Math.round((2000 - rating) / 100 + 1);
            this.rankClass = RankClass.KYU;
        } else {
            this.rank = Math.round((rating - 2000) / 100);
            this.rankClass = RankClass.DAN;
        }
    }

    public format() {
        if (this.rankClass === RankClass.KYU) {
            return `${this.rank} Kyu`;
        }
        return `${this.rank} Dan`;
    }

    public formatShort() {
        if (this.rankClass === RankClass.KYU) {
            return `${this.rank}k`;
        }
        return `${this.rank}d`;
    }

    public equals(other: Rank) {
        return this.rank === other.rank && this.rankClass === other.rankClass;
    }

    public greaterThan(other: Rank) {
        if (this.rankClass === other.rankClass) {
            return this.rankClass === RankClass.KYU ? this.rank < other.rank : this.rank > other.rank;
        }
        if (this.rankClass === RankClass.DAN) {
            return true;
        }
        return false;
    }
}

export function formatRank(rating: number) {
    const rank = new Rank(rating);
    return rank.format();
}

export function formatRankShort(rating: number) {
    const rank = new Rank(rating);
    return rank.formatShort();
}
