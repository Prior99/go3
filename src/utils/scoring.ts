export enum GameResult {
    WIN = "win",
    LOSS = "loss",
    TIE = "tie",
}

export enum RankClass {
    KYU = "kyu",
    DAN = "dan",
}

export interface Rank {
    rank: number;
    rankClass: RankClass;
}

/**
 * Implementation of the official EGF rating system.
 * http://www.europeangodatabase.eu/EGD/EGF_rating_system.php
 */
const epsilon = 0.016;

function winningExpectancy(ratingDifference: number) {
    // SE(A) = 1 / (e ^ (D / a) + 1) - epsilon / 2
    return 1 / (Math.exp(ratingDifference / adjustment(ratingDifference)) + 1) - epsilon / 2;
}

function adjustment(rating: number) {
    return 205 - Math.floor(rating / 20);
}

function magnitude(rating: number) {
    return Math.round(122 - 6 * (rating / 100) + Math.pow(rating / 100, 2) / 15);
}

function scoreFromResult(result: GameResult) {
    switch (result) {
        case GameResult.WIN: return 1;
        case GameResult.TIE: return 0.5;
        case GameResult.LOSS: return 0;
        default: return;
    }
}

export function newRating(oldRating: number, otherRating: number, result: GameResult) {
    const actualScore = scoreFromResult(result);
    const expectedScore = winningExpectancy(otherRating - oldRating);
    let difference = actualScore - expectedScore;
    return Math.round(oldRating + magnitude(oldRating) * (actualScore - expectedScore));
}

export function rankFromRating(rating: number): Rank {
    if (rating < 2100) {
        return {
            rank: Math.round((2000 - rating) / 100 + 1),
            rankClass: RankClass.KYU,
        };
    }
    return {
        rank: Math.round((rating - 2000) / 100),
        rankClass: RankClass.DAN,
    };
}

export function formatRank(rating: number) {
    const rank = rankFromRating(rating);
    if (rank.rankClass === RankClass.KYU) {
        return `${rank.rank} Kyu`;
    }
    return `${rank.rank} Dan`;
}

export function formatRankShort(rating: number) {
    const rank = rankFromRating(rating);
    if (rank.rankClass === RankClass.KYU) {
        return `${rank.rank}k`;
    }
    return `${rank.rank}d`;
}
