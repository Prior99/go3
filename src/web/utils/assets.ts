import { component } from "tsdi";
import { bindAll } from "lodash-decorators";
import * as downscale from "downscale";

function getKey(url: string, width?: number, height?: number) {
    if (typeof width !== "undefined" && typeof height !== "undefined") {
        return `${width}__${height}__${url}`;
    }
    return url;
}

@component
@bindAll()
export class Assets {
    private images = new Map<string, HTMLImageElement>();
    private loading = new Map<string, ((HTMLImageElement) => void)[]>();
    private scaling = new Map<string, ((HTMLImageElement) => void)[]>();

    public get loaded() { return this.loading.size === 0; }

    public loadImage(url: string, width?: number, height?: number) {
        if (typeof width === "undefined" || typeof height === "undefined") {
            return this.downloadImage(url);
        }
        return new Promise<HTMLImageElement>(async resolve => {
            const key = getKey(url, width, height);
            if (this.images.has(key)) {
                return resolve(this.images.get(key));
            }
            if (!this.scaling.has(key)) {
                this.scaling.set(key, []);
                const original = await this.downloadImage(url);
                const scaled = await downscale(original, width, height, { imageType: "png" });
                const scaledImg = new Image();
                scaledImg.addEventListener("load", () => {
                    this.images.set(key, scaledImg);
                    this.scaling.get(key).forEach(callback => callback(scaledImg));
                    this.scaling.delete(key);
                });
                scaledImg.src = scaled;
            }
            this.scaling.get(key).push(resolve);
        });
    }

    private downloadImage(url: string) {
        return new Promise<HTMLImageElement>(resolve => {
            if (this.images.has(url)) {
                return resolve(this.images.get(url));
            }
            if (!this.loading.has(url)) {
                this.loading.set(url, []);
                const img = new Image();
                img.addEventListener("load", () => {
                    this.images.set(url, img);
                    this.loading.get(url).forEach(callback => callback(img));
                    this.loading.delete(url);
                });
                img.src = url;
            }
            this.loading.get(url).push(resolve);
        });
    }

    public get(url: string, width?: number, height?: number) {
        return this.images.get(getKey(url, width, height));
    }
}
