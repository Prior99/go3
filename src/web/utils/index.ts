export * from "./assets";
export * from "./rendering";
