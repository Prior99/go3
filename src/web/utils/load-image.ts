import { component } from "tsdi";
import { bind } from "decko";

@component
export class Assets {
    private images = new Map<string, HTMLImageElement>();
    private loading = new Map<string, ((HTMLImageElement) => void)[]>;

    public get loaded() { return loading.length === 0; }

    @bind public loadImage(url: string) {
        return new Promise<HTMLImageElement>(resolve => {
            if (!this.loading.has(url)) {
                this.loading.set(url, []);
            }
            this.loading.get(url).push(resolve);
            const img = new Image();
            img.src = url;
            img.addEventListener("load", () => {
                this.images.set(url, img);
                this.loading.get(url).forEach(callback => callback(img));
                this.loading.delete(url);
            });
        });
    }

    @bind public get(url: string) {
        return this.images.get(url);
    }
}
