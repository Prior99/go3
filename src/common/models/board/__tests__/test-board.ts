import { testBoard } from "./utils";
import { Board } from "../";
import { Color } from "../../../utils";

[
    {
        board: testBoard(
            "WBB " +
            "WW B" +
            "WB  " +
            "WB  ",
        ),
        pos: { col: 1, row: 1 },
    },
    {
        board: testBoard(
            "  W " +
            " BWW" +
            " BW " +
            "  W ",
        ),
        pos: { col: 2, row: 0 },
    },
    {
        board: testBoard(
            "    " +
            "W W " +
            " W  " +
            "W W ",
        ),
        pos: { col: 1, row: 2 },
    },
    {
        board: testBoard(
            "WWWW" +
            "W  W" +
            "W  W" +
            "WWWW",
        ),
        pos: { col: 0, row: 0 },
    },
].forEach(({ board, pos }) => {
    test("detects the groups correctly", () => {
        expect(board.groupAt(board.toIndex(pos)).indices.map(board.toPos)).toMatchSnapshot();
    });
});

test("constructing a board from an old one", () => {
    debugger
    const board = testBoard(
        "WWWB " +
        "W WB " +
        "WWWB " +
        "BBBB " +
        "     ",
    );
    (board as any).id = "some-id";
    board.prisonersBlack = 10;
    board.prisonersWhite = 10;
    board.created = new Date("2017-11-15 12:00:00");
    board.updated = new Date("2017-11-15 12:00:00");
    board.turn = 6;

    const newBoard = new Board(board, board.toIndex({ col: 1, row: 1 }));
    expect(newBoard).toMatchSnapshot();
});

[
    { turn: 0, color: Color.BLACK },
    { turn: 1, color: Color.WHITE },
    { turn: 2, color: Color.BLACK },
].forEach(({ turn, color }) => {
    test(`color ${color} is to play at turn ${turn}`, () => {
        const board = testBoard("    ");
        board.turn = turn;
        expect(board.currentColor).toEqual(color);
    });
});

[
    {
        board: testBoard(
            "B " +
            "  ",
        ),
        equal: testBoard(
            "B " +
            "  ",
        ),
        other: testBoard(
            "  " +
            "  ",
        ),
    },
    {
        board: testBoard(
            "WWW" +
            "W  " +
            "W  ",
        ),
        equal: testBoard(
            "WWW" +
            "W  " +
            "W  ",
        ),
        other: testBoard(
            "WWW" +
            "   " +
            "W  ",
        ),
    },
].forEach(({ board, equal, other }) => {
    test(`boards compared for equality`, () => {
        expect(board.equals(equal)).toBe(true);
        expect(board.equals(other)).toBe(false);
    });
});

test("converting indices to positions and back", () => {
    const board = testBoard(
        "   " +
        "   " +
        "   ",
    );
    expect(board.toPos(0)).toEqual({ col: 0, row: 0 });
    expect(board.toPos(4)).toEqual({ col: 1, row: 1 });
    expect(board.toPos(6)).toEqual({ col: 0, row: 2 });
    expect(board.toIndex({ col: 0, row: 0 })).toEqual(0);
    expect(board.toIndex({ col: 1, row: 1 })).toEqual(4);
    expect(board.toIndex({ col: 0, row: 2 })).toEqual(6);
});

test("getting the neighbours of position", () => {
    const board = testBoard(
        "   " +
        " W " +
        "   ",
    );
    expect(board.neighbours(board.toIndex({ col: 1, row: 1 })).map(board.toPos)).toMatchSnapshot();
    expect(board.neighbours(board.toIndex({ col: 0, row: 0 })).map(board.toPos)).toMatchSnapshot();
    expect(board.neighbours(board.toIndex({ col: 1, row: 0 })).map(board.toPos)).toMatchSnapshot();
    expect(board.neighbours(board.toIndex({ col: 2, row: 2 })).map(board.toPos)).toMatchSnapshot();
    expect(board.neighbours(board.toIndex({ col: 1, row: 2 })).map(board.toPos)).toMatchSnapshot();
});

test("getting the neighbours of position with specified color", () => {
    const board = testBoard(
        " B " +
        "W W" +
        " B ",
    );
    expect(board.neighboursOfColor(board.toIndex({ col: 1, row: 1 }), Color.BLACK).map(board.toPos))
        .toMatchSnapshot();
    expect(board.neighboursOfColor(board.toIndex({ col: 0, row: 0 }), Color.BLACK).map(board.toPos))
        .toMatchSnapshot();
    expect(board.neighboursOfColor(board.toIndex({ col: 1, row: 0 }), Color.BLACK).map(board.toPos))
        .toMatchSnapshot();
    expect(board.neighboursOfColor(board.toIndex({ col: 2, row: 2 }), Color.WHITE).map(board.toPos))
        .toMatchSnapshot();
    expect(board.neighboursOfColor(board.toIndex({ col: 1, row: 2 }), Color.BLACK).map(board.toPos))
        .toMatchSnapshot();
});

test("getting the neighbours of position with the same color", () => {
    const board = testBoard(
        " B " +
        "WWW" +
        " B ",
    );
    expect(board.neighboursOfSameColor(board.toIndex({ col: 1, row: 1 })).map(board.toPos)).toMatchSnapshot();
});

test("placing a token", () => {
    const board = testBoard(
        "BW   " +
        "BWWW " +
        "BW W " +
        "BWWW " +
        "B    ",
    );
    (board as any).id = "some-id";
    board.prisonersBlack = 0;
    board.prisonersWhite = 0;
    board.created = new Date("2017-11-15 12:00:00");
    board.updated = new Date("2017-11-15 12:00:00");
    board.turn = 5;

    const newBoard = board.place(board.toIndex({ col: 1, row: 4 }));
    expect(newBoard).toMatchSnapshot();
});

test("passing", () => {
    const board = testBoard(
        "BW " +
        "BWW" +
        " W ",
    );
    (board as any).id = "some-id";
    board.created = new Date("2017-11-15 12:00:00");
    board.updated = new Date("2017-11-15 12:00:00");
    board.turn = 5;

    const newBoard = board.pass();
    expect(newBoard).toMatchSnapshot();
});

test("resigning", () => {
    const board = testBoard(
        " W " +
        "WWW" +
        " W ",
    );
    (board as any).id = "some-id";
    board.created = new Date("2017-11-15 12:00:00");
    board.updated = new Date("2017-11-15 12:00:00");
    board.turn = 5;

    const newBoard = board.resign();
    expect(newBoard).toMatchSnapshot();
});

test("placing a token without triggering game logics", () => {
    const board = testBoard(
        "BW   " +
        "BWWW " +
        "BW W " +
        "BWWW " +
        "B    ",
    );
    (board as any).id = "some-id";
    board.prisonersBlack = 0;
    board.prisonersWhite = 0;
    board.created = new Date("2017-11-15 12:00:00");
    board.updated = new Date("2017-11-15 12:00:00");
    board.turn = 5;

    const newBoard = board.mockPlace(board.toIndex({ col: 1, row: 4 }));
    expect(newBoard).toMatchSnapshot();
});

test("getting the prisoners by color", () => {
    const board = testBoard("    ");
    board.prisonersBlack = 1;
    board.prisonersWhite = 2;
    expect(board.prisoners(Color.BLACK)).toEqual(1);
    expect(board.prisoners(Color.WHITE)).toEqual(2);
});
