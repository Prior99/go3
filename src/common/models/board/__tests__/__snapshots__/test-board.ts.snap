// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructing a board from an old one 1`] = `
Board {
  "game": undefined,
  "groupCache": Map {
    7 => Group {
      "board": [Circular],
      "indices": Array [
        7,
        2,
        1,
        0,
        5,
        10,
        11,
        12,
      ],
    },
    11 => Group {
      "board": [Circular],
      "indices": Array [
        11,
        10,
        5,
        0,
        1,
        2,
        7,
        12,
      ],
    },
    5 => Group {
      "board": [Circular],
      "indices": Array [
        5,
        0,
        1,
        2,
        7,
        12,
        11,
        10,
      ],
    },
    1 => Group {
      "board": [Circular],
      "indices": Array [
        1,
        0,
        5,
        10,
        11,
        12,
        7,
        2,
      ],
    },
  },
  "parent": Board {
    "created": 2017-11-15T12:00:00.000Z,
    "groupCache": Map {},
    "id": "some-id",
    "prisonersBlack": 10,
    "prisonersWhite": 10,
    "state": Array [
      "white",
      "white",
      "white",
      "black",
      "empty",
      "white",
      "empty",
      "white",
      "black",
      "empty",
      "white",
      "white",
      "white",
      "black",
      "empty",
      "black",
      "black",
      "black",
      "black",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
      "empty",
    ],
    "turn": 6,
    "updated": 2017-11-15T12:00:00.000Z,
  },
  "placedAt": 6,
  "prisonersBlack": 18,
  "prisonersWhite": 10,
  "state": Array [
    "empty",
    "empty",
    "empty",
    "black",
    "empty",
    "empty",
    "black",
    "empty",
    "black",
    "empty",
    "empty",
    "empty",
    "empty",
    "black",
    "empty",
    "black",
    "black",
    "black",
    "black",
    "empty",
    "empty",
    "empty",
    "empty",
    "empty",
    "empty",
  ],
  "turn": 7,
}
`;

exports[`detects the groups correctly 1`] = `
Array [
  Object {
    "col": 1,
    "row": 1,
  },
  Object {
    "col": 0,
    "row": 1,
  },
  Object {
    "col": 0,
    "row": 0,
  },
  Object {
    "col": 0,
    "row": 2,
  },
  Object {
    "col": 0,
    "row": 3,
  },
]
`;

exports[`detects the groups correctly 2`] = `
Array [
  Object {
    "col": 2,
    "row": 0,
  },
  Object {
    "col": 2,
    "row": 1,
  },
  Object {
    "col": 2,
    "row": 2,
  },
  Object {
    "col": 2,
    "row": 3,
  },
  Object {
    "col": 3,
    "row": 1,
  },
]
`;

exports[`detects the groups correctly 3`] = `
Array [
  Object {
    "col": 1,
    "row": 2,
  },
]
`;

exports[`detects the groups correctly 4`] = `
Array [
  Object {
    "col": 0,
    "row": 0,
  },
  Object {
    "col": 0,
    "row": 1,
  },
  Object {
    "col": 0,
    "row": 2,
  },
  Object {
    "col": 0,
    "row": 3,
  },
  Object {
    "col": 1,
    "row": 3,
  },
  Object {
    "col": 2,
    "row": 3,
  },
  Object {
    "col": 3,
    "row": 3,
  },
  Object {
    "col": 3,
    "row": 2,
  },
  Object {
    "col": 3,
    "row": 1,
  },
  Object {
    "col": 3,
    "row": 0,
  },
  Object {
    "col": 2,
    "row": 0,
  },
  Object {
    "col": 1,
    "row": 0,
  },
]
`;

exports[`getting the neighbours of position 1`] = `
Array [
  Object {
    "col": 2,
    "row": 1,
  },
  Object {
    "col": 1,
    "row": 2,
  },
  Object {
    "col": 0,
    "row": 1,
  },
  Object {
    "col": 1,
    "row": 0,
  },
]
`;

exports[`getting the neighbours of position 2`] = `
Array [
  Object {
    "col": 1,
    "row": 0,
  },
  Object {
    "col": 0,
    "row": 1,
  },
]
`;

exports[`getting the neighbours of position 3`] = `
Array [
  Object {
    "col": 2,
    "row": 0,
  },
  Object {
    "col": 1,
    "row": 1,
  },
  Object {
    "col": 0,
    "row": 0,
  },
]
`;

exports[`getting the neighbours of position 4`] = `
Array [
  Object {
    "col": 1,
    "row": 2,
  },
  Object {
    "col": 2,
    "row": 1,
  },
]
`;

exports[`getting the neighbours of position 5`] = `
Array [
  Object {
    "col": 2,
    "row": 2,
  },
  Object {
    "col": 0,
    "row": 2,
  },
  Object {
    "col": 1,
    "row": 1,
  },
]
`;

exports[`getting the neighbours of position with specified color 1`] = `
Array [
  Object {
    "col": 1,
    "row": 2,
  },
  Object {
    "col": 1,
    "row": 0,
  },
]
`;

exports[`getting the neighbours of position with specified color 2`] = `
Array [
  Object {
    "col": 1,
    "row": 0,
  },
]
`;

exports[`getting the neighbours of position with specified color 3`] = `Array []`;

exports[`getting the neighbours of position with specified color 4`] = `
Array [
  Object {
    "col": 2,
    "row": 1,
  },
]
`;

exports[`getting the neighbours of position with specified color 5`] = `Array []`;

exports[`getting the neighbours of position with the same color 1`] = `
Array [
  Object {
    "col": 2,
    "row": 1,
  },
  Object {
    "col": 0,
    "row": 1,
  },
]
`;

exports[`passing 1`] = `
Board {
  "game": undefined,
  "groupCache": Map {},
  "parent": Board {
    "created": 2017-11-15T12:00:00.000Z,
    "groupCache": Map {},
    "id": "some-id",
    "state": Array [
      "black",
      "white",
      "empty",
      "black",
      "white",
      "white",
      "empty",
      "white",
      "empty",
    ],
    "turn": 5,
    "updated": 2017-11-15T12:00:00.000Z,
  },
  "prisonersBlack": undefined,
  "prisonersWhite": undefined,
  "state": Array [
    "black",
    "white",
    "empty",
    "black",
    "white",
    "white",
    "empty",
    "white",
    "empty",
  ],
  "turn": 6,
}
`;

exports[`placing a token 1`] = `
Board {
  "game": undefined,
  "groupCache": Map {
    20 => Group {
      "board": [Circular],
      "indices": Array [
        20,
        15,
        10,
        5,
        0,
      ],
    },
  },
  "parent": Board {
    "created": 2017-11-15T12:00:00.000Z,
    "groupCache": Map {},
    "id": "some-id",
    "prisonersBlack": 0,
    "prisonersWhite": 0,
    "state": Array [
      "black",
      "white",
      "empty",
      "empty",
      "empty",
      "black",
      "white",
      "white",
      "white",
      "empty",
      "black",
      "white",
      "empty",
      "white",
      "empty",
      "black",
      "white",
      "white",
      "white",
      "empty",
      "black",
      "empty",
      "empty",
      "empty",
      "empty",
    ],
    "turn": 5,
    "updated": 2017-11-15T12:00:00.000Z,
  },
  "placedAt": 21,
  "prisonersBlack": 0,
  "prisonersWhite": 5,
  "state": Array [
    "empty",
    "white",
    "empty",
    "empty",
    "empty",
    "empty",
    "white",
    "white",
    "white",
    "empty",
    "empty",
    "white",
    "empty",
    "white",
    "empty",
    "empty",
    "white",
    "white",
    "white",
    "empty",
    "empty",
    "white",
    "empty",
    "empty",
    "empty",
  ],
  "turn": 6,
}
`;

exports[`placing a token without triggering game logics 1`] = `
Board {
  "game": undefined,
  "groupCache": Map {},
  "parent": Board {
    "created": 2017-11-15T12:00:00.000Z,
    "groupCache": Map {},
    "id": "some-id",
    "prisonersBlack": 0,
    "prisonersWhite": 0,
    "state": Array [
      "black",
      "white",
      "empty",
      "empty",
      "empty",
      "black",
      "white",
      "white",
      "white",
      "empty",
      "black",
      "white",
      "empty",
      "white",
      "empty",
      "black",
      "white",
      "white",
      "white",
      "empty",
      "black",
      "empty",
      "empty",
      "empty",
      "empty",
    ],
    "turn": 5,
    "updated": 2017-11-15T12:00:00.000Z,
  },
  "prisonersBlack": 0,
  "prisonersWhite": 0,
  "state": Array [
    "black",
    "white",
    "empty",
    "empty",
    "empty",
    "black",
    "white",
    "white",
    "white",
    "empty",
    "black",
    "white",
    "empty",
    "white",
    "empty",
    "black",
    "white",
    "white",
    "white",
    "empty",
    "black",
    "white",
    "empty",
    "empty",
    "empty",
  ],
  "turn": 6,
}
`;

exports[`resigning 1`] = `
Board {
  "game": undefined,
  "groupCache": Map {},
  "parent": Board {
    "created": 2017-11-15T12:00:00.000Z,
    "groupCache": Map {},
    "id": "some-id",
    "state": Array [
      "empty",
      "white",
      "empty",
      "white",
      "white",
      "white",
      "empty",
      "white",
      "empty",
    ],
    "turn": 5,
    "updated": 2017-11-15T12:00:00.000Z,
  },
  "prisonersBlack": undefined,
  "prisonersWhite": undefined,
  "resigned": true,
  "state": Array [
    "empty",
    "white",
    "empty",
    "white",
    "white",
    "white",
    "empty",
    "white",
    "empty",
  ],
  "turn": 6,
}
`;
